projects/com.oracle.truffle.llvm.parser.impl/src/com/oracle/truffle/llvm/parser/impl/LLVMVisitor.java

```diff
@@ -869,21 +869,21 @@ private LLVMExpressionNode getElementPtr(LLVMExpressionNode baseAddress, Type ba
                 LLVMExpressionNode valueRef = visitValueRef(currentRef, type);
                 currentAddress = factoryFacade.createGetElementPtr(getLLVMType(type), currentAddress, valueRef, indexedTypeLength);
             } else {
-                int indexedTypeLength = LLVMTypeHelper.goIntoTypeGetLengthByte(currentType, constantIndex);
+                int stride = LLVMTypeHelper.getByteSize(currentType.getContainedType(0));
                 currentType = LLVMTypeHelper.goIntoType(currentType, constantIndex);
-                if (indexedTypeLength != 0) {
+                if (stride != 0) {
                     LLVMExpressionNode constantNode;
                     switch (getLLVMType(type)) {
                         case I32:
-                            constantNode = factoryFacade.createLiteral(1, LLVMBaseType.I32);
+                            constantNode = factoryFacade.createLiteral(constantIndex.intValue(), LLVMBaseType.I32);
                             break;
                         case I64:
-                            constantNode = factoryFacade.createLiteral(1L, LLVMBaseType.I64);
+                            constantNode = factoryFacade.createLiteral(constantIndex.longValue(), LLVMBaseType.I64);
                             break;
                         default:
                             throw new AssertionError();
                     }
-                    currentAddress = factoryFacade.createGetElementPtr(getLLVMType(type), currentAddress, constantNode, indexedTypeLength);
+                    currentAddress = factoryFacade.createGetElementPtr(getLLVMType(type), currentAddress, constantNode, stride);
                 }
```

Comment:
The change in `LLVMVisitor` ignores that the GEP instruction can also index structs since it only takes in account the stride computed by the first element and the value of the GEP index. `LLVMTypeHelper.goIntoTypeGetLengthByte` computed the offset into a struct or array. Maybe you can introduce a new method in LLVMTypeHelper that returns a tuple (stride, index) that returns the the stride and index for arrays, but uses `LLVMTypeHelper.goIntoTypeGetLengthByte` and constant 1 otherwise?